File: Numeros Pseudo-Aleatorios/SplitMix64/splitmix64.py
"""
   PRNG .......: SplitMix64
   Ano ........: 2015
   Autor ......: Sebastiano Vigna
"""

# Fundamentado: rosettacode.org/wiki/Pseudo-random_numbers/Splitmix64

import time


class SplitMix64:
    def __init__(self, state):
        self.state = state

    def next_int(self):
        self.state += 0x9e3779b97f4a7c15
        z = self.state
        z = (z ^ (z >> 30)) * 0xbf58476d1ce4e5b9
        z = (z ^ (z >> 27)) * 0x94d049bb133111eb
        return z & 0xFFFFFFFFFFFFFFFF  # Limita a 64 bits

    def next_float(self):
        return float(self.next_int() >> 11) / (1 << 53)


Xn1 = SplitMix64(1234)
Xn2 = SplitMix64(int(time.time() * 1e9))

print("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)")
print("PSEUDORANDOM NUMBER GENERATOR (PRNG)")
print("PADRAO: SplitMix64")
print("===================================================")
print("")

print("  Semente: 1234 (int)        Semente: TIME (int)")
print("  --------------------       --------------------")
for i in range(1, 6):
    print(f"{Xn1.next_int():22}{Xn2.next_int():27}")

print("\n  Semente: 1234 (real)       Semente: TIME (real)")
print("  --------------------       --------------------")
for i in range(1, 6):
    print(f"{Xn1.next_float():22.18f}{Xn2.next_float():27.18f}")

File: Numeros Pseudo-Aleatorios/SplitMix64/splitmix64.go
/*
 PRNG .......: SplitMix64
 Ano ........: 2015
 Autor ......: Sebastiano Vigna
*/

// Fundamentado: rosettacode.org/wiki/Pseudo-random_numbers/Splitmix64

package main

import (
	"fmt"
	"time"
)

type SplitMix64 struct {
	state uint64
}

func SplitMix64Init(state uint64) SplitMix64 {
	return SplitMix64{state}
}

func (sm64 *SplitMix64) nextInt() uint64 {
	sm64.state += 0x9e3779b97f4a7c15
	z := sm64.state
	z = (z ^ (z >> 30)) * 0xbf58476d1ce4e5b9
	z = (z ^ (z >> 27)) * 0x94d049bb133111eb
	return z ^ (z >> 31)
}

func (sm64 *SplitMix64) nextFloat() float64 {
	return float64(sm64.nextInt()>>11) / (1 << 53)
}

func main() {
	
	Xn1 := SplitMix64Init(1234)
	Xn2 := SplitMix64Init(uint64(time.Now().UnixNano()))

	fmt.Println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)")
	fmt.Println("PSEUDORANDOM NUMBER GENERATOR (PRNG)")
	fmt.Println("PADRAO: SplitMix64")
	fmt.Println("===================================================\n")

	fmt.Println("  Semente: 1234 (int)        Semente: TIME (int)")
	fmt.Println("  --------------------       --------------------")
	for i := 1; i <= 5; i++ {
		fmt.Printf("%22d%27d\n", Xn1.nextInt(), Xn2.nextInt())
	}
	fmt.Println()

	fmt.Println("  Semente: 1234 (real)       Semente: TIME (real)")
	fmt.Println("  --------------------       --------------------")
	for i := 1; i <= 5; i++ {
		fmt.Printf("%22.18f%27.18f\n", Xn1.nextFloat(), Xn2.nextFloat())
	}
	
}
